hibernate-core/src/main/java/org/hibernate/boot/model/internal/PropertyInferredData.java

```diff
@@ -88,23 +90,41 @@ public TypeDetails getPropertyType() throws MappingException {
 					sourceModelContext.getClassDetailsRegistry(),
 					() -> new DynamicClassDetails( targetName, sourceModelContext )
 			);
-			return new ClassTypeDetailsImpl( classDetails, TypeDetails.Kind.CLASS );
+			return new ClassTypeDetailsImpl( classDetails, Kind.CLASS );
 		}
 
-		final Target legacyTargetAnnotation = propertyMember.getDirectAnnotationUsage( Target.class );
-		if ( legacyTargetAnnotation != null ) {
-			return resolveLegacyTargetAnnotation( legacyTargetAnnotation, sourceModelContext );
+		final TargetEmbeddable targetEmbeddable = getTargetEmbeddableAnnotation(propertyMember);
+		if ( targetEmbeddable != null ) {
+			return resolveTargetEmbeddableAnnotation( targetEmbeddable, sourceModelContext );
 		}
 
 		return propertyMember.resolveRelativeType( ownerType );
 	}
 
-	private static ClassTypeDetailsImpl resolveLegacyTargetAnnotation(
-			Target legacyTargetAnnotation,
+	private TargetEmbeddable getTargetEmbeddableAnnotation(MemberDetails memberDetails) {
+		TargetEmbeddable targetEmbeddable = memberDetails.getDirectAnnotationUsage( TargetEmbeddable.class );
+		if ( targetEmbeddable != null ) {
+			if (!memberDetails.hasDirectAnnotationUsage(Embedded.class) ) {
+				// If set on an attribute, it should be an @Embedded
+				throw new MappingException( "The @TargetEmbeddable annotation can only be set on an element marked with @Embedded" );
+			}
+		}
+		else {
+			// Set on an interface
+			final TypeDetails typeDetails = memberDetails.getAssociatedType();
+			targetEmbeddable = typeDetails.getTypeKind() != Kind.CLASS ?
+					null :
+					typeDetails.asClassType().getClassDetails().getDirectAnnotationUsage( TargetEmbeddable.class );
+		}
```

Comment:
@sebersole had to change this wrt the previous implementation to get it to work for interfaces; pls check this is correct.

Reply:
```suggestion
			targetEmbeddable = memberDetails
					.getAssociatedType()
					.determineRawClass()
					.getDirectAnnotationUsage( TargetEmbeddable.class );
		}
```